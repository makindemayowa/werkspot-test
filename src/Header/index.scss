.header {
    margin-right: auto;
    margin-left: auto;
    padding: 0.7em 1em;
    border-bottom: thin solid #dce1e4;
    .flex {
      display: flex;
      justify-content: space-between;
      align-items: center;
    }
    .item-cont {
      position: relative;
      .item {
        display: flex;
        align-items: center;
        color: #454444;
        padding: 0.3em;
        font-size: 14px;
        cursor: pointer;
  
        &.input {
          border: thin solid #b2b6b2;
          padding: 0 0.3em;
          border-radius: 0.2em;
  
          .icon {
            margin-right: 0.1em;
          }
  
          input {
            border: none;
            padding: 0.3em 0;
  
            &:focus {
              outline: none;
            }
          }
        }
      }
    }
  }
  