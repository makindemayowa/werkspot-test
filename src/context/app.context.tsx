import * as React from 'react';
import { ITodoData } from '../types';

interface IState {
  toDos: ITodoData[];
  searchString: string
};

interface IActionCreator {
  type: string;
  data: any;
};

type IDispatch = ({ type, data }: IActionCreator) => null | void;

type IReducer = (state: IState, action: IActionCreator) => IState;

interface IActionType {
  ADD_NEW_CARD: string;
  SET_TODOS: string;
  SET_SEARCH: string;
  UPDATE_CARD_COLUMN: string;
};

const getTodos = (): ITodoData[] => {
    const todoString = localStorage.getItem('toDos');
    let todos: ITodoData[] = [];
    if(todoString) {
        todos = JSON.parse(todoString);
    };
    return todos;
}

const setTodos = (data: ITodoData[]) => localStorage.setItem('toDos', JSON.stringify(data));

const initialState: IState = {
  toDos: getTodos(),
  searchString: ''
};

export const actionType: IActionType = {
  ADD_NEW_CARD: 'ADD_NEW_CARD',
  UPDATE_CARD_COLUMN: 'UPDATE_CARD_COLUMN',
  SET_TODOS: 'SET_TODOS',
  SET_SEARCH: 'SET_SEARCH'
};

const reducer: IReducer = (state, action) => {
  switch (action.type) {
    case actionType.SET_TODOS:
      setTodos(action.data)
      return {
        ...state,
        toDos: action.data,
      };
    case actionType.UPDATE_CARD_COLUMN:
      const toDosCopy = Array.from(state.toDos);
      const cardToUpdate = toDosCopy[action.data.originalIdx].rows[action.data.cardIdx]
      toDosCopy[action.data.originalIdx].rows.splice(action.data.cardIdx, 1)
      toDosCopy[action.data.nextIdx].rows.unshift(cardToUpdate)
      setTodos(toDosCopy)
      return {
        ...state,
        toDos: toDosCopy,
      };
    case actionType.ADD_NEW_CARD:
      const cloned = Array.from(state.toDos);
      const itemIndex = action.data.columnIndex;
      const updatedColumn = cloned[itemIndex];
      updatedColumn.rows.push(action.data.newCard);
      cloned[itemIndex] = updatedColumn;
      setTodos(cloned)
      return {
        ...state,
        toDos: cloned
      };
    case actionType.SET_SEARCH:
      return {
        ...state,
        searchString: action.data
      };
    default:
      return state;
  }
};

export const AppStateContext = React.createContext<IState | undefined>(initialState);
export const AppDispatchContext = React.createContext<IDispatch | undefined>(
  undefined
);

export function useAppState() {
  const state = React.useContext(AppStateContext);

  if (state === undefined) {
    throw new Error('You can only use useAppState inside a context provider');
  }

  return state;
};

export function useAppDispatch() {
  const dispatch = React.useContext(AppDispatchContext);

  if (dispatch === undefined) {
    throw new Error(
      'You can only use useAppDispatch inside a context provider'
    );
  };

  return dispatch;
};

export function AppProvider({ children }: { children: any }) {
  const [state, dispatch] = React.useReducer(reducer, initialState);

  return (
    <AppStateContext.Provider value={state}>
      <AppDispatchContext.Provider value={dispatch}>
        {children}
      </AppDispatchContext.Provider>
    </AppStateContext.Provider>
  );
};
