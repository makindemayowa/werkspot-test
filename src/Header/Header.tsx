import * as React from 'react'
import './index.scss';
import removeRedEyeIcon from '../images/remove-red-eye.svg';
import searchIcon from '../images/search.svg';
import { actionType, useAppDispatch } from '../context/app.context';

export default function Header() {
  const [searchTerm, setSearchTerm] = React.useState<string>('');
  const dispatch = useAppDispatch()

  const onSearchtermChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(event.target.value);
  };

  React.useEffect(() => {
    dispatch({ type: actionType.SET_SEARCH, data: searchTerm });
  }, [searchTerm, dispatch])

  return (
    <div className="container-fluid header">
      <div className="flex">
        <div>
          <div className="item-cont">
            <div className="item">
                <img src={removeRedEyeIcon} alt="remove red eye icon" />
                <span>Statuses</span>
            </div>
          </div>
        </div>
        <div>
          <div className="flex">
            <div className="item-cont">
              <div className="item input">
                <img src={searchIcon} alt="search icon" />
                <input
                  value={searchTerm}
                  onChange={onSearchtermChange}
                  type="text"
                  placeholder="Search by title"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}
