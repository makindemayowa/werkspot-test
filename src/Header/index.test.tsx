import enzyme from 'enzyme'
import renderer from 'react-test-renderer';
import { AppProvider } from '../context/app.context';
import Header from './Header';

test('Header element creates a consistent snapshot', () => {
    const component = renderer.create(<AppProvider><Header /></AppProvider>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('Should contain an input element', () => {
  const wrapper = enzyme.render(<AppProvider><Header /></AppProvider>);
  expect(wrapper.find('input')).toBeDefined();
});