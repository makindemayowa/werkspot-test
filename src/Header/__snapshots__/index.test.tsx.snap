// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header element creates a consistent snapshot 1`] = `
<div
  className="container-fluid header"
>
  <div
    className="flex"
  >
    <div>
      <div
        className="item-cont"
      >
        <div
          className="item"
        >
          <img
            alt="remove red eye icon"
            src="remove-red-eye.svg"
          />
          <span>
            Statuses
          </span>
        </div>
      </div>
    </div>
    <div>
      <div
        className="flex"
      >
        <div
          className="item-cont"
        >
          <div
            className="item input"
          >
            <img
              alt="search icon"
              src="search.svg"
            />
            <input
              onChange={[Function]}
              placeholder="Search by title"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
